File: MarkerPassingAlgorithm/src/main/java/de/kimanufaktur/markerpassing/Node.java
package de.kimanufaktur.markerpassing;

import java.util.Collection;

/**
 * The node represents the vertex to the edges in the graph. A node holds the markers passed to it. It can be the target or the source oa a link.
 */
public interface Node {

    //Link for the network created out of nodes
    default void addLink(Link link) {
        getLinks().add(link);
    }

    default void removeLink(Link link) {
        getLinks().remove(link);
    }

    Collection<Link> getLinks();

    //de.kimanufaktur.markerpassing.Marker holding the activation information of the node
    default void addMarker(Marker marker) {
        getMarkers().add(marker);
    }

    default void removeMarker(Marker marker) {
        getMarkers().remove(marker);
    }

    Collection<Marker> getMarkers();

    //de.kimanufaktur.markerpassing.Node management functions

    /**
     * Check the Threshold of the node. If the threshold is exceeded, the node is added to the active nodes and can
     * be selected for firing.
     *
     * @param markerClasses the markers to check the threshold for.
     * @return boolean if one of the threshold is reached
     */
    boolean checkThresholds(Object markerClasses);
}

File: marker_passing/node.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Collection, TYPE_CHECKING

if TYPE_CHECKING:
    from .link import Link
    from .marker import Marker


class Node(ABC):
    """
    Vertex in the graph. Holds markers and is connected via links.
    """

    def add_link(self, link: "Link") -> None:
        self.get_links().append(link)

    def remove_link(self, link: "Link") -> None:
        self.get_links().remove(link)

    @abstractmethod
    def get_links(self) -> list["Link"]:
        ...

    def add_marker(self, marker: "Marker") -> None:
        self.get_markers().append(marker)

    def remove_marker(self, marker: "Marker") -> None:
        self.get_markers().remove(marker)

    @abstractmethod
    def get_markers(self) -> list["Marker"]:
        ...

    @abstractmethod
    def check_thresholds(self, marker_classes: object) -> bool:
        """
        Returns True if any threshold is reached for the given marker classes,
        meaning this node should be considered active.
        """
        ...
